jax_distribution/systems/ippo_ff_anakin.py

```diff
@@ -0,0 +1,475 @@
+        (
+            new_params,
+            new_opt_state,
+            rng,
+            new_env_state,
+            new_timestep,
+            metric_info,
+        ) = update_step_fn(
+            params,
+            opt_state,
+            rng,
+            env_state,
+            timestep,
+        )
+
+        return (
+            new_params,
+            new_opt_state,
+            rng,
+            new_env_state,
+            new_timestep,
+        ), metric_info
+
+    def learner_fn(params, opt_state, rngs, env_states, timesteps):
+        """Vectorise and repeat the update."""
+        batched_update_fn = jax.vmap(
+            update_fn, axis_name="j"
+        )  # vectorize across batch.
+
+        def iterate_fn(val, _):  # repeat many times to avoid going back to Python.
+            params, opt_state, rngs, env_states, timesteps = val
+            return batched_update_fn(params, opt_state, rngs, env_states, timesteps)
+
+        runner_state = (params, opt_state, rngs, env_states, timesteps)
+        runner_state, metric = jax.lax.scan(
+            iterate_fn, runner_state, None, config["ITERATIONS"]
+        )
+        return {"runner_state": runner_state, "metrics": metric}
+
+    return learner_fn
+
+
+def run_experiment(env_name, config):
+    if "MultiCVRP" in env_name:
+        config["NUM_AGENTS"] = 3
+        generator = UniformRandomGenerator(
+            num_vehicles=config["NUM_AGENTS"], num_customers=6
+        )
+        env = jumanji.make(env_name, generator=generator)
+        num_actions = int(env.action_spec().maximum)
+
+    elif "RobotWarehouse" in env_name:
+        env = jumanji.make(env_name)
+        num_actions = int(env.action_spec().num_values[0])
+        config["NUM_AGENTS"] = env.num_agents
+
+    env = LogWrapper(env)
+    cores_count = len(jax.devices())
+    # Number of iterations
+    timesteps_per_iteration = (
+        cores_count * config["ROLLOUT_LENGTH"] * config["BATCH_SIZE"]
+    )
+    config["ITERATIONS"] = (
+        config["TOTAL_TIMESTEPS"] // timesteps_per_iteration
+    )  # Number of training updates
+
+    num_frames = config["TOTAL_TIMESTEPS"]
+
+    network = ActorCritic(
+        num_actions, activation=config["ACTIVATION"], env_name=env_name
+    )
+    optim = optax.chain(
+        optax.clip_by_global_norm(config["MAX_GRAD_NORM"]),
+        optax.adam(config["LR"], eps=1e-5),
+    )
+
+    rng = jax.random.PRNGKey(config["SEED"])
+    rng, rng_env, rng_params = jax.random.split(rng, 3)
+
+    init_obs = env.observation_spec().generate_value()
+    init_obs = jax.tree_util.tree_map(
+        lambda x: x[None, ...],
+        init_obs,
+    )
+
+    params = network.init(rng_params, init_obs)
+    opt_state = optim.init(params)
+
+    # TODO: Complete this
+    learn = get_learner_fn(
+        env,
+        network.apply,
+        optim.update,
+        config,
+    )
+
+    learn = jax.pmap(learn, axis_name="i")  # replicate over multiple cores.
```

Comment:
Another nice chnage will be `"i"` -> `"devices"` to show the axis.

Reply:
Done! Also do you have a better name for `"j"`?

Reply:
Maybe we can use "i" -> "device" and "j" -> "batch".

Reply:
sounds great :)